lib/AST/Decl.cpp

```diff
@@ -1165,7 +1165,13 @@ static bool isPolymorphic(const AbstractStorageDecl *storage) {
 
   case DeclKind::Class:
     // Final properties can always be direct, even in classes.
-    return !storage->isFinal();
+    if (storage->isFinal())
+      return false;
+    // Extension properties are statically dispatched, unless they're @objc.
+    if (storage->getDeclContext()->isExtensionContext()
+        && !storage->isObjC())
```

Comment:
This isn't keypath-specific, it's part of `AbstractStorageDecl::getAccessStrategy`. When we change the language, this seems like the right place to centralize that logic.